checker/src/org/checkerframework/checker/index/upperbound/UpperBoundStore.java

```diff
@@ -0,0 +1,337 @@
+import org.checkerframework.framework.flow.CFAbstractStore;
+import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.source.Result;
+import org.checkerframework.framework.type.AnnotatedTypeFactory;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.javacutil.AnnotationUtils;
+import org.checkerframework.javacutil.ElementUtils;
+
+public class UpperBoundStore extends CFAbstractStore<CFValue, UpperBoundStore> {
+
+    private static /*@CompilerMessageKey*/ String NO_REASSIGN = "reassignment.not.permitted";
+    private static /*@CompilerMessageKey*/ String NO_REASSIGN_FIELD =
+            "reassignment.field.not.permitted";
+    private static /*@CompilerMessageKey*/ String SIDE_EFFECTING_METHOD =
+            "side.effect.invalidation";
+    private static /*@CompilerMessageKey*/ String NO_REASSIGN_FIELD_METHOD =
+            "reassignment.field.not.permitted.method";
+
+    private enum CheckController {
+        NAME_ONLY,
+        NAME_AND_METHOD_CALLS,
+        NONFINAL_REFS_AND_METHOD_CALLS,
+        NONFINAL_REFS_AND_METHOD_CALLS_METHOD
+    }
+
+    public UpperBoundStore(UpperBoundAnalysis analysis, boolean sequentialSemantics) {
+        super(analysis, sequentialSemantics);
+    }
+
+    public UpperBoundStore(
+            UpperBoundAnalysis analysis, CFAbstractStore<CFValue, UpperBoundStore> other) {
+        super(other);
+    }
+
+    @Override
+    public void updateForMethodCall(
+            MethodInvocationNode n, AnnotatedTypeFactory factory, CFValue val) {
+        super.updateForMethodCall(n, factory, val);
+
+        ExecutableElement elt = n.getTarget().getMethod();
+
+        if (!isSideEffectFree(factory, elt)) {
+            List<? extends Element> enclosedElts =
+                    ElementUtils.enclosingClass(elt).getEnclosedElements();
+            List<AnnotatedTypeMirror> enclosedTypes = new ArrayList<AnnotatedTypeMirror>();
+
+            findEnclosedTypes(enclosedTypes, enclosedElts);
+
+            // Should include all method calls and non-array references, but not the name of the method.
+            clearFromStore("", n, CheckController.NONFINAL_REFS_AND_METHOD_CALLS_METHOD);
+            checkForRemainingAnnotations(
+                    enclosedTypes, "", n, CheckController.NONFINAL_REFS_AND_METHOD_CALLS_METHOD);
+        }
+    }
+
+    @Override
+    public void updateForAssignment(Node n, CFValue val) {
+        // Do reassignment things here.
+        super.updateForAssignment(n, val);
+
+        // This code determines the list of dependences in types that are to be invalidated
+        CheckController checkController = null;
+
+        if (n.getType().getKind() == TypeKind.ARRAY) {
+            if (n instanceof LocalVariableNode) {
+                // Do not warn about assigning to a final variable. javac handles this.
+                if (!ElementUtils.isFinal(((LocalVariableNode) n).getElement())) {
```

Comment:
Yes. These should all be calls to `isEffectivelyFinal`, and I'm not sure how I screwed that up.